Add Data Valuation Issue Manager

coding-famer opened the pull request:
For issue #740

elisno commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,135 @@
+    description: ClassVar[
+        str
+    ] = """Examples who won't contribute a lot for a model's training
+     are flagged as having label issues.
```
Wait, should we compare this issue with label issues?

coding-famer replied:
sorry this is a mistake

elisno commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,135 @@
+    ):
+        super().__init__(datalab)
+        self.k = k
+        if threshold is None:
+            self.threshold = self.DEFAULT_THRESHOLDS
+        else:
+            self.threshold = threshold
```
You can set this with a ternary operator.

```suggestion
        self.thresholdd = threshold if threshold is not None else self.DEFAULT_THRESHOLDS
```

Apply this to similar if-else blocks.

elisno commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,135 @@
+        dist = knn_graph.indices.reshape(N, -1)
+        self.k = dist.shape[1]
+
+        for i, y in enumerate(labels):
+            idx = dist[i][::-1]
+            ans = labels[idx]
+            s[idx[self.k - 1]][i] = float(ans[self.k - 1] == y) / self.k
+            cur = self.k - 2
+            for j in range(self.k - 1):
+                s[idx[cur]][i] = s[idx[cur + 1]][i] + float(
+                    int(ans[cur] == y) - int(ans[cur + 1] == y)
+                ) / self.k * (min(cur, self.k - 1) + 1) / (cur + 1)
+                cur -= 1
```
Consider extracting this loop (and related variables if necessary) to another internal method and use more descriptive variable names.

Not always apparent what each line is doing.

elisno reviewed (commented):
Thank you for the PR @coding-famer!

I made a quick, initial pass through the `data_valuation.py` module.
I'll follow up with a more thorough review of the remainder of the PR shortly.

smttsp commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,137 @@
+    ):
+        super().__init__(datalab)
+        self.k = k
+        self.threshold = threshold if threshold is not None else self.DEFAULT_THRESHOLDS
```
a formatting tool, like `black` would convert this to:

```suggestion
        self.threshold = (
             threshold 
             if threshold is not None 
             else self.DEFAULT_THRESHOLDS
         )
```

jwmueller replied:
@smttsp Excited to see you want to help contribute to this library!  

I suggest first reading through [our developer guide](https://github.com/cleanlab/cleanlab/blob/master/DEVELOPMENT.md) to learn all of our conventions and see what is happening in our CI. You'll see we are already using `black`, `mypy`, and many other things you suggested. They just don't happen to be in this PR, and if you look at [the CI logs](https://github.com/cleanlab/cleanlab/actions/runs/6194792316/job/16818336127?pr=850), you'll see it is complaining about these things already (automatically so reviewers don't have to)

jwmueller replied:
@smttsp If you'd like to contribute code, a good place to get started is with a GH issue marked as 'good first issue'

https://github.com/cleanlab/cleanlab/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22

smttsp replied:
@jwmueller, thanks for your response. Happy to contribute! 

Sure, I read the developer guide and a bunch of other stuff. It is quite large and great work!

> good first issue

The issues seem to be quite complex and require some good understanding of the repo and the work. But I will play with the repo today.

smttsp commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,137 @@
+        knn_graph_stats = self.datalab.get_info("statistics").get("weighted_knn_graph", None)
+
+        knn_graph: Optional[csr_matrix] = None
+        if knn_graph_kwargs is not None:
+            knn_graph = knn_graph_kwargs
+        elif knn_graph_stats is not None:
+            knn_graph = knn_graph_stats
```
this is essentially equal to 

```
    knn_graph = knn_graph_kwargs or knn_graph_stats
```

I can also see why you wanna write it more verbose, but just a suggestion

jwmueller replied:
I prefer the more verbose version for clarity

smttsp commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,137 @@
+        elif knn_graph_stats is not None:
+            knn_graph = knn_graph_stats
+
+        if isinstance(knn_graph, csr_matrix) and kwargs.get("k", 0) > (
+            knn_graph.nnz // knn_graph.shape[0]
+        ):
```
if (
        isinstance(knn_graph, csr_matrix) 
        and kwargs.get("k", 0) > (knn_graph.nnz // knn_graph.shape[0])
    ):
        knn_graph = None

this may be a little more clear and easier to understand

smttsp commented on cleanlab/datalab/internal/issue_manager/data_valuation.py:
```diff
@@ -0,0 +1,137 @@
+            "average_data_valuation": issues[self.issue_score_key].mean(),
+        }
+
+        info_dict = {
+            **issues_info,
+        }
```
is info_dict not the same as issues_info?

smttsp reviewed (commented):
I like the work you guys are doing, and I have seen your repo from reddit. Wanted to contribute a bit. 

I have a few suggestions below. One thing I think is missing is a formatting tool. 

I would suggest 

`black`: general formatting
`isort`: formatting and reordering imports
`ruff`: linter
`mypy`: type hints

please feel free to disregard the suggestions if you think they are not relevant

jwmueller commented:
@coding-famer Note that your PR is failing the CI check for Format (run `black` to fix that as mentioned in DEVELOPMENT.md).

The other failing CI checks (type check and link check) will be auto-resolved if you merge in the latest master branch (we've fixed these since).

coding-famer commented:
> @coding-famer Note that your PR is failing the CI check for Format (run `black` to fix that as mentioned in DEVELOPMENT.md).
> 
> The other failing CI checks (type check and link check) will be auto-resolved if you merge in the latest master branch (we've fixed these since).

Hi, it seems that the format issue isn't related to my added file:

> would reformat /home/runner/work/cleanlab/cleanlab/tests/datalab/test_factory.py
> 
> Oh no! 💥 💔 💥
> 1 file would be reformatted, 112 files would be left unchanged.

Do I need to reformat it too?

jwmueller commented:
> > @coding-famer Note that your PR is failing the CI check for Format (run `black` to fix that as mentioned in DEVELOPMENT.md).
> > The other failing CI checks (type check and link check) will be auto-resolved if you merge in the latest master branch (we've fixed these since).
> 
> Hi @jwmueller , it seems that the format issue isn't related to my added file:
> 
> > would reformat /home/runner/work/cleanlab/cleanlab/tests/datalab/test_factory.py
> > Oh no! 💥 💔 💥
> > 1 file would be reformatted, 112 files would be left unchanged.
> 
> Do I need to reformat it too?

Didn't you edit that part? The part `black` is complaining about says: "data_valuation" and is this line of the code:
https://github.com/cleanlab/cleanlab/blob/fd127726399ba8affafa9ee7070aed47e77a1f61/tests/datalab/test_factory.py#L16

jwmueller commented:
@elisno please confirm this new issue type is NOT part of the set of issue-types that Datalab checks for by default. At first, this should only be accessible to users who specifically request it (especially until we have a better notion of it's robustness and additional runtime overheads).

coding-famer commented:
> > > @coding-famer Note that your PR is failing the CI check for Format (run `black` to fix that as mentioned in DEVELOPMENT.md).
> > > The other failing CI checks (type check and link check) will be auto-resolved if you merge in the latest master branch (we've fixed these since).
> > 
> > 
> > Hi @jwmueller , it seems that the format issue isn't related to my added file:
> > > would reformat /home/runner/work/cleanlab/cleanlab/tests/datalab/test_factory.py
> > > Oh no! 💥 💔 💥
> > > 1 file would be reformatted, 112 files would be left unchanged.
> > 
> > 
> > Do I need to reformat it too?
> 
> Didn't you edit that part? The part `black` is complaining about says: "data_valuation" and is this line of the code:
> 
> https://github.com/cleanlab/cleanlab/blob/fd127726399ba8affafa9ee7070aed47e77a1f61/tests/datalab/test_factory.py#L16

Sorry my fault, updated.

smttsp commented on tests/datalab/test_factory.py:
```diff
@@ -13,7 +13,14 @@ def registry():
     issue_types = Datalab.list_possible_issue_types()
     assert isinstance(issue_types, list)
-    possible_issues = ["label", "outlier", "near_duplicate", "non_iid", "class_imbalance"]
+    possible_issues = [
```
you might wanna have this as a `pytest fixture`.

smttsp replied:
something like this:
```
@pytest.fixture
def possible_issues():
    return ["label", "outlier", "near_duplicate", "non_iid", "class_imbalance"]
```

coding-famer replied:
Hi, I think this is not related to this PR, maybe you can create a new PR to improve the pytest?

smttsp replied:
I see, I thought this was a new unit test. 

I am not really sure what these codes are doing, so I would prefer not to contribute without having a good understanding of the code, to be honest. 

you might wanna add a `#TODO` if you think this would take long time.

smttsp reviewed (commented):
one suggestion

elisno reviewed (commented):
This looks great so far!

- At this point, we just need some clearer docs on how the knn-shapley score fits in for the data valuation issue manger. Could you add a a docstring for the relevant methods to highlight its importance in this issue manager.
- Can you add some end-to-end tests where this issue manager is run via Datalab?

coding-famer commented:
@elisno 
Thanks for your comment!
For 
> Can you add some end-to-end tests where this issue manager is run via Datalab?

Do you mean add a test with `datalab.find_issue(...)`?
If do, where can I add this test case, in the `test_datalab`? And what do we expect to see?

elisno commented:
> Do you mean add a test with `datalab.find_issue(...)`? If do, where can I add this test case, in the `test_datalab`? And what do we expect to see?

Yes, just add a new test class in test_datalab.py that's specific to the use of this issue manager via Datalab.

We typically play with different configurations of the standard inputs (`pred_probs`, ` features`, `knn_graph`, etc.).

In this case, it looks like we're just starting with support for knn_graphs. I can see several end-to-end tests we want to cover in the beginning:

1. Support a pre-computed knn_graph.
2. Test the effects of passing in features and not a knn_graph.
3. Test that it reuses the state properly (like if we incrementally run Datalab, first with an issue_type like "outlier" to store a `knn_graph`, then the second run we run the `data_valuation` issue_type and somehow test that the same knn_graph is used.

I think the 3rd point is the easiest to get started with. The other two tests require more work.

elisno reviewed (approved):
Excellent work @coding-famer!
